File: crates/inspect-api/src/prompts.rs
use inspect_core::detect::types::DetectorFinding;
use inspect_core::types::EntityReview;
use sem_core::model::change::ChangeType;

/// DEPRECATED: Use `SYSTEM_VALIDATE_FINDINGS` for the new deterministic pipeline.
/// Kept for backward compatibility during migration.
pub const SYSTEM_REVIEW: &str = "You are a precise code reviewer. Only report real bugs you are confident about. Always respond with valid JSON.";

pub const SYSTEM_VALIDATE: &str = "You are a precise reviewer. Verify each issue against the actual diff. Only keep confirmed bugs. Always respond with valid JSON.";

pub const SYSTEM_VALIDATE_FINDINGS: &str = r#"You are a code review validation gate.

<instructions>
- You receive static analysis findings with evidence and entity metadata.
- Your ONLY job is to classify each finding as true_positive or false_positive by checking the actual diff.
- Do NOT invent new issues. Do NOT expand scope beyond the listed findings.
- Never fabricate line numbers, variable names, or code snippets not present in the diff.
- When uncertain whether a finding is real, default to false_positive. Precision over recall.
- Keep explanations to one sentence per verdict.
- Always respond with valid JSON matching the exact schema requested.
</instructions>"#;

/// DEPRECATED: Use `PROMPT_VALIDATE_FINDINGS` via `format_validate_findings_prompt` for the new
/// deterministic pipeline. Kept for backward compatibility during migration.
pub const PROMPT_DEEP: &str = r#"You are a world-class code reviewer. Review this PR and find ONLY real, concrete bugs.

PR Title: {pr_title}

{triage_section}

PR Diff:
{diff}

Look specifically for these categories of issues:
1. Logic errors: wrong conditions, off-by-one, incorrect algorithms, broken control flow, inverted booleans
2. Concurrency bugs: race conditions, missing locks, unsafe shared state, deadlocks, unhandled async promises
3. Null/undefined safety: missing null checks, possible NPE, Optional.get() without isPresent(), uninitialized variables
4. Error handling: swallowed exceptions, missing error propagation, wrong error types
5. Data correctness: wrong translations, wrong constants, incorrect mappings, copy-paste errors, stale cache data
6. Security: SSRF, XSS, injection, auth bypass, exposed secrets, unsafe deserialization, origin validation bypass
7. Type mismatches: wrong return types, incompatible casts, API contract violations, schema errors
8. Breaking changes: removed public APIs without migration, changed behavior silently
9. State consistency: asymmetric cache trust, orphaned data, inconsistent updates across related fields
10. Naming/contract bugs: method name typos that break interfaces, property names that don't match expected contracts

Rules:
- ONLY report issues you are highly confident about (>90% sure)
- Be specific: name the file, function/variable, and exactly what's wrong
- Naming typos ARE bugs if they would cause a runtime error or break an API contract
- Do NOT report: style preferences, missing tests, docs, "could be improved"
- Do NOT report issues about code that was only deleted/removed
- Maximum 10 issues. Quality over quantity.

For each issue, provide it as a JSON object with "issue" (description) and "evidence" (quote the specific code lines from the diff that prove this is a bug).

Respond with ONLY a JSON object:
{{"issues": [{{"issue": "description", "evidence": "the specific code"}}]}}"#;

pub const PROMPT_VALIDATE_FINDINGS: &str = r#"# Task
Validate static analysis findings against the PR diff. For each finding, determine true_positive or false_positive.

# Context
PR Title: {pr_title}

{triage_section}

# PR Diff
{diff}

# Findings to validate
{findings_list}

# Reasoning steps (for each finding)
1. Locate the evidence snippet in the diff. If not found → false_positive.
2. Read the entity source code provided. Check if the pattern match is coincidental (e.g., a variable named "password" in a password-validator is not a hardcoded secret).
3. Consider the metadata: a finding on a public API with high blast_radius affects more code. A high risk_score means more dependents.
4. Decide: would this cause incorrect behavior, a crash, or a security issue in production?

# Rules
<scope_constraints>
- Only validate the findings listed above. Do NOT invent new issues.
- Do NOT expand scope or suggest improvements beyond the listed findings.
- When in doubt, mark false_positive. Precision over recall.
- You MUST cite the exact code from the diff that confirms a true_positive.
- Never fabricate line numbers, variable names, or code that is not in the diff.
</scope_constraints>

# Self-check
Before finalizing your response, re-scan each true_positive verdict:
- Is the cited evidence actually present in the diff above?
- Could this be a false alarm from a naming coincidence or test/example code?
If either answer is yes, downgrade to false_positive.

# Output format
<output_spec>
- Respond with ONLY a JSON object, no markdown fences, no commentary.
- One verdict per finding. Explanation: one sentence max with a code citation.
</output_spec>
{{"verdicts": [{{"rule_id": "...", "entity_name": "...", "verdict": "true_positive" | "false_positive", "explanation": "one sentence with code citation"}}]}}"#;

pub const PROMPT_DEEP_FALLBACK: &str = r#"# Task
The deterministic analyzers found no issues, but entity-level triage flagged high-risk changes. Review only for high-confidence bugs.

# Context
PR Title: {pr_title}

{triage_section}

# PR Diff
{diff}

# Bug categories to check
1. Null/undefined dereference that WILL crash at runtime
2. Security vulnerabilities (SSRF, XSS, SQL injection, auth bypass)
3. Concurrency bugs (race conditions, missing locks)
4. Logic errors that are provably wrong (wrong operator, inverted condition)

# Rules
<scope_constraints>
- Maximum 3 issues. Only report if >95% confident.
- You MUST quote the exact code line as evidence.
- Pay attention to entity metadata: focus on entities marked [PUBLIC API] or with high blast_radius.
- Do NOT report style, naming, missing tests, or "could be improved" issues.
- Do NOT invent issues not supported by the diff. If the diff is clean, return an empty issues array.
- Never fabricate line numbers or code snippets not present in the diff.
</scope_constraints>

# Self-check
Before finalizing, verify each issue:
- Is the quoted evidence actually in the diff above?
- Is the confidence truly >95%, or is this speculative?
If either fails, drop the issue.

# Output format
<output_spec>
- Respond with ONLY a JSON object, no markdown fences, no commentary.
- Maximum 3 issues. Each must include evidence quoting the exact code line.
</output_spec>
{{"issues": [{{"issue": "description", "evidence": "exact code line", "severity": "critical|high"}}]}}"#;

pub const PROMPT_VALIDATE: &str = r#"You are a senior code reviewer doing final validation. You have the PR diff and candidate issues.

PR Title: {pr_title}

PR Diff (for verification):
{diff}

Candidate Issues:
{candidates}

For each candidate, verify against the actual diff:
1. Can you find the specific code that's buggy? If yes, keep it.
2. Is this a real bug that would cause incorrect behavior in production? If yes, keep it.
3. Is this about deleted/removed code being replaced? If so, DROP it.
4. Is this speculative or theoretical ("could potentially...")? If so, DROP it.
5. Is this about style, naming conventions, or missing tests? If so, DROP it.

Return ONLY the issues that are verified real bugs with evidence in the diff.

Respond with ONLY a JSON object:
{{"issues": ["verified issue 1", "verified issue 2", ...]}}"#;

/// Smart diff truncation that deprioritizes tests, docs, configs.
pub fn truncate_diff(diff: &str, max_chars: usize) -> String {
    if diff.len() <= max_chars {
        return diff.to_string();
    }

    let parts: Vec<&str> = diff.split("diff --git ").collect();
    if parts.is_empty() {
        return diff[..max_chars].to_string();
    }

    let mut scored: Vec<(f64, &str)> = Vec::new();
    for part in &parts {
        let trimmed = part.trim();
        if trimmed.is_empty() {
            continue;
        }

        let adds = part.matches("\n+").count().saturating_sub(part.matches("\n+++").count());
        let dels = part.matches("\n-").count().saturating_sub(part.matches("\n---").count());
        let mod_bonus = adds.min(dels) * 2;
        let mut score = (adds + dels + mod_bonus) as f64;

        let first_line = part.lines().next().unwrap_or("").to_lowercase();

        // Deprioritize test files
        if ["test", "spec", "mock", "__test__", "fixture"]
            .iter()
            .any(|kw| first_line.contains(kw))
        {
            score *= 0.3;
        }
        // Deprioritize docs
        if [".md", ".adoc", ".txt", ".rst", "changelog", "readme"]
            .iter()
            .any(|kw| first_line.contains(kw))
        {
            score *= 0.2;
        }
        // Deprioritize snapshots/lockfiles
        if [".snap", ".lock", "package-lock", "yarn.lock"]
            .iter()
            .any(|kw| first_line.contains(kw))
        {
            score *= 0.1;
        }
        // Deprioritize config files
        if [".json", ".yaml", ".yml", ".toml", ".xml"]
            .iter()
            .any(|kw| first_line.contains(kw))
        {
            score *= 0.5;
        }

        scored.push((score, part));
    }

    scored.sort_by(|a, b| b.0.partial_cmp(&a.0).unwrap_or(std::cmp::Ordering::Equal));

    let mut result = String::new();
    for (_, part) in &scored {
        let candidate = format!("diff --git {}", part);
        if result.len() + candidate.len() > max_chars {
            break;
        }
        result.push_str(&candidate);
    }

    if result.is_empty() {
        diff[..max_chars].to_string()
    } else {
        result
    }
}

/// Build entity-grouped triage section from entity reviews.
/// Includes risk_score, blast_radius, dependent_count, and public API status.
pub fn build_rich_triage(entities: &[EntityReview]) -> String {
    if entities.is_empty() {
        return String::new();
    }

    let mut meaningful: Vec<&EntityReview> = entities
        .iter()
        .filter(|e| {
            matches!(
                e.change_type,
                ChangeType::Modified | ChangeType::Added
            ) && e.entity_type != "chunk"
        })
        .collect();

    meaningful.sort_by(|a, b| b.risk_score.partial_cmp(&a.risk_score).unwrap_or(std::cmp::Ordering::Equal));
    let top: Vec<&EntityReview> = meaningful.into_iter().take(20).collect();

    if top.is_empty() {
        return String::new();
    }

    // Group by file
    let mut by_file: std::collections::HashMap<&str, Vec<&EntityReview>> =
        std::collections::HashMap::new();
    for e in &top {
        by_file.entry(e.file_path.as_str()).or_default().push(e);
    }

    let mut file_entries: Vec<(&str, Vec<&EntityReview>)> = by_file.into_iter().collect();
    file_entries.sort_by(|a, b| {
        let a_max = a.1.iter().map(|e| e.risk_score).fold(0.0_f64, f64::max);
        let b_max = b.1.iter().map(|e| e.risk_score).fold(0.0_f64, f64::max);
        b_max.partial_cmp(&a_max).unwrap_or(std::cmp::Ordering::Equal)
    });

    let mut lines = vec!["## Entity-level triage (highest-risk changes)".to_string()];
    lines.push("Each entity includes: name, type, change_type, classification, risk (score), blast_radius, dependents count.".to_string());
    for (fp, ents) in &file_entries {
        lines.push(format!("\n### {}", fp));
        for e in ents {
            let public = if e.is_public_api { " [PUBLIC API]" } else { "" };
            lines.push(format!(
                "- `{}` ({}, {:?}, {}) | risk={} ({:.2}) | blast_radius={} | dependents={}{public}",
                e.entity_name,
                e.entity_type,
                e.change_type,
                e.classification,
                e.risk_level,
                e.risk_score,
                e.blast_radius,
                e.dependent_count,
            ));
        }
    }

    lines.join("\n")
}

/// DEPRECATED: Use format_validate_findings_prompt for the new deterministic pipeline.
/// Kept for backward compatibility during migration.
pub fn format_deep_prompt(pr_title: &str, triage_section: &str, diff: &str) -> String {
    PROMPT_DEEP
        .replace("{pr_title}", pr_title)
        .replace("{triage_section}", triage_section)
        .replace("{diff}", diff)
}

/// Format the PROMPT_VALIDATE template with actual values.
pub fn format_validate_prompt(pr_title: &str, diff: &str, candidates: &str) -> String {
    PROMPT_VALIDATE
        .replace("{pr_title}", pr_title)
        .replace("{diff}", diff)
        .replace("{candidates}", candidates)
}

/// Format the PROMPT_VALIDATE_FINDINGS template with deterministic findings and entity context.
pub fn format_validate_findings_prompt(
    pr_title: &str,
    triage_section: &str,
    diff: &str,
    findings: &[DetectorFinding],
    entity_reviews: &[EntityReview],
) -> String {
    let findings_list: String = findings
        .iter()
        .enumerate()
        .map(|(i, f)| {
            // Find the matching entity review to include source context
            let entity = entity_reviews
                .iter()
                .find(|e| e.entity_id == f.entity_id);

            let mut entry = format!(
                "<finding id=\"{}\">\nrule_id: {}\nseverity: {} | confidence: {:.0}%\nentity: `{}` in {}\nmessage: {}\nevidence: `{}`",
                i + 1,
                f.rule_id,
                f.severity,
                f.confidence * 100.0,
                f.entity_name,
                f.file_path,
                f.message,
                f.evidence,
            );

            if let Some(e) = entity {
                entry.push_str(&format!(
                    "\nrisk: {} ({:.2}) | blast_radius: {} | dependents: {}{}",
                    e.risk_level,
                    e.risk_score,
                    e.blast_radius,
                    e.dependent_count,
                    if e.is_public_api { " | PUBLIC API" } else { "" },
                ));
                // Include entity after_content (capped) so the LLM can read the full function
                if let Some(ref content) = e.after_content {
                    let capped: String = content.chars().take(1500).collect();
                    entry.push_str(&format!("\nentity_source:\n```\n{}\n```", capped));
                }
            }

            entry.push_str("\n</finding>");
            entry
        })
        .collect::<Vec<_>>()
        .join("\n\n");

    PROMPT_VALIDATE_FINDINGS
        .replace("{pr_title}", pr_title)
        .replace("{triage_section}", triage_section)
        .replace("{diff}", diff)
        .replace("{findings_list}", &findings_list)
}

/// Format the PROMPT_DEEP_FALLBACK template for high-risk changes with no deterministic findings.
pub fn format_deep_fallback_prompt(pr_title: &str, triage_section: &str, diff: &str) -> String {
    PROMPT_DEEP_FALLBACK
        .replace("{pr_title}", pr_title)
        .replace("{triage_section}", triage_section)
        .replace("{diff}", diff)
}

File: agent/src/prompts.ts
import { diffLines } from "diff";
import type { DetectorFinding, EntityReview } from "./types.js";

/** System prompt: review protocol, not just a bug list. */
export const SYSTEM_PROMPT_CLAUDE = `You are a precision code reviewer. Find only high-confidence, concrete correctness bugs in changed code.

Review protocol:
1. FIRST: analyze all code snippets provided below WITHOUT using tools. Identify concrete suspicions before any tool use.
2. For every method shown (especially @Override, public API, or interface implementations), check:
   - Contract check: does the implementation satisfy what the method name, interface, Javadoc, and callers expect?
   - Return-value check: does it return the correct type/object, not a related but wrong one? (e.g., getX() must return X, not Y)
   - Fluent/builder check: are return values from builder/fluent APIs captured, or silently discarded (making the call a no-op)?
   - Dead code check: is any computed result unused or overwritten?
   - Guard check: are safety checks (null guards, bounds checks, assertions) present where needed?
3. Use tools ONLY to verify a specific suspicion — never for broad exploration.

Do NOT report: style, naming, missing tests, documentation, suggestions, or issues in deleted-only code.

Respond with ONLY a JSON object:
{"issues": [{"issue": "description naming exact function/variable and the concrete bug", "evidence": "exact code snippet", "severity": "critical|high|medium", "file": "path/to/file"}]}

Return {"issues": []} if no bugs.`;

export const SYSTEM_PROMPT_GPT = SYSTEM_PROMPT_CLAUDE;

export const SYSTEM_PROMPT = SYSTEM_PROMPT_CLAUDE;

/**
 * Build a focused user prompt from the top-N ranked entities.
 *
 * Key design decisions:
 * - Top-30 entities by risk_score (covers 127/128 golden bugs at top-20, +10 buffer)
 * - No file expansion — only ranked entities, preserving pipeline signal
 * - Contract co-location: for @Override methods, inline the interface signature/Javadoc
 * - Collapse repeated boilerplate (e.g., multiple UnsupportedOperationException stubs)
 */
export function buildUserPrompt(
  prTitle: string,
  _diff: string,
  _triageSection: string,
  findings: DetectorFinding[],
  entityReviews: EntityReview[],
): string {
  const TOP_N = 30;
  const topEntities = entityReviews.slice(0, TOP_N);

  // Build a lookup of ALL entities by name for contract co-location
  const entitiesByName = new Map<string, EntityReview[]>();
  for (const e of entityReviews) {
    const arr = entitiesByName.get(e.entity_name) ?? [];
    arr.push(e);
    entitiesByName.set(e.entity_name, arr);
  }

  // Detect which entities are interface/abstract declarations (contracts)
  const topEntityIds = new Set(topEntities.map((e) => e.entity_id));

  // Group top entities by file for display
  const byFile = new Map<string, EntityReview[]>();
  for (const e of topEntities) {
    const arr = byFile.get(e.file_path) ?? [];
    arr.push(e);
    byFile.set(e.file_path, arr);
  }
  const sortedFiles = [...byFile.entries()].sort((a, b) => {
    const aMax = Math.max(...a[1].map((e) => e.risk_score));
    const bMax = Math.max(...b[1].map((e) => e.risk_score));
    return bMax - aMax;
  });

  const lines: string[] = [];
  lines.push(`# PR: ${prTitle || "(untitled)"}`);
  lines.push(`Top ${topEntities.length} entities across ${sortedFiles.length} files (from ${entityReviews.length} total)`);
  lines.push("");

  // Findings — strongest signals first
  if (findings.length > 0) {
    lines.push(`## Detector Findings (${findings.length})`);
    for (const f of findings) {
      const entity = entityReviews.find((e) => e.entity_id === f.entity_id);
      const risk = entity
        ? ` | risk=${entity.risk_level}(${entity.risk_score.toFixed(2)}) deps=${entity.dependent_count}`
        : "";
      lines.push(
        `- **[${f.severity.toUpperCase()}] ${f.rule_id}** \`${f.entity_name}\` ${f.file_path}:${f.start_line}`,
      );
      lines.push(`  ${f.message}`);
      lines.push(`  \`${f.evidence}\`${risk}`);
    }
    lines.push("");
  }

  // Per-file entity sections
  for (const [filePath, entities] of sortedFiles) {
    const maxLevel = entities.reduce((best, e) =>
      e.risk_score > (best?.risk_score ?? 0) ? e : best,
    ).risk_level;

    lines.push(`## ${filePath} (${maxLevel}, ${entities.length} entities)`);

    // Collapse UnsupportedOperationException stubs
    const stubs: EntityReview[] = [];
    const nonStubs: EntityReview[] = [];
    for (const e of entities) {
      const content = e.after_content ?? "";
      if (
        e.change_type === "Added" &&
        content.includes("throw new UnsupportedOperationException(") &&
        content.length < 400
      ) {
        stubs.push(e);
      } else {
        nonStubs.push(e);
      }
    }

    if (stubs.length > 1) {
      const stubNames = stubs.map((s) => `\`${s.entity_name}\``).join(", ");
      const maxDeps = Math.max(...stubs.map((s) => s.dependent_count));
      lines.push(
        `**${stubs.length} stub methods** (throw UnsupportedOperationException): ${stubNames} — max ${maxDeps} callers each`,
      );
      lines.push("");
    }

    for (const e of stubs.length > 1 ? nonStubs : entities) {
      const pub = e.is_public_api ? " **PUBLIC**" : "";
      const callers = e.dependent_names
        .slice(0, 4)
        .map(([n]) => n)
        .join(", ");
      const callerStr = callers
        ? ` | callers: ${callers}${e.dependent_names.length > 4 ? ` +${e.dependent_names.length - 4}` : ""}`
        : "";

      lines.push(
        `### \`${e.entity_name}\` (${e.entity_type}, ${e.change_type}) :${e.start_line}-${e.end_line}`,
      );
      lines.push(
        `risk=${e.risk_level}(${e.risk_score.toFixed(2)}) blast=${e.blast_radius} deps=${e.dependent_count}${pub}${callerStr}`,
      );

      // Contract co-location: for @Override methods, find the interface/parent signature
      const after = e.after_content ?? "";
      if (after.includes("@Override") && e.entity_type === "method") {
        const contract = findContract(e, entityReviews, entitiesByName);
        if (contract) {
          lines.push("**Contract:**");
          lines.push("```");
          lines.push(contract);
          lines.push("```");
        }
      }

      const before = e.before_content ?? "";

      if (before && after && e.change_type === "Modified") {
        const changes = diffLines(before, after);
        const diffStr: string[] = [];
        for (const part of changes) {
          const prefix = part.added ? "+" : part.removed ? "-" : " ";
          const partLines = part.value.split("\n").filter((l) => l !== "");
          if (prefix === " " && partLines.length > 6) {
            diffStr.push(" " + partLines.slice(0, 2).join("\n "));
            diffStr.push(`  ... (${partLines.length - 4} unchanged lines)`);
            diffStr.push(" " + partLines.slice(-2).join("\n "));
          } else {
            for (const line of partLines) {
              diffStr.push(prefix + line);
            }
          }
        }
        lines.push("```diff");
        lines.push(diffStr.join("\n"));
        lines.push("```");
      } else if (after) {
        if (after.length <= 800) {
          lines.push("```");
          lines.push(after);
          lines.push("```");
        } else {
          lines.push("```");
          lines.push(after.slice(0, 600));
          lines.push(`... (${after.length - 600} more chars)`);
          lines.push("```");
        }
      } else if (before) {
        if (before.length <= 800) {
          lines.push("```");
          lines.push(before);
          lines.push("```");
        } else {
          lines.push("```");
          lines.push(before.slice(0, 600));
          lines.push(`... (${before.length - 600} more chars — deleted)`);
          lines.push("```");
        }
      }
      lines.push("");
    }
  }

  return lines.join("\n");
}

/**
 * Find the interface/parent contract for an @Override method.
 * Looks for an entity with the same name in a different file that is an
 * interface method or has Javadoc describing the contract.
 */
function findContract(
  entity: EntityReview,
  allEntities: EntityReview[],
  byName: Map<string, EntityReview[]>,
): string | null {
  // Strategy 1: look for same-named method entity that is an interface/abstract method
  const candidates = byName.get(entity.entity_name) ?? [];
  for (const c of candidates) {
    if (c.entity_id === entity.entity_id) continue;
    if (c.file_path === entity.file_path) continue;
    const content = c.after_content ?? c.before_content ?? "";
    if (!content) continue;
    // Pure interface method (no body, just signature) — best contract source
    if (!content.includes("{") || content.includes("interface ")) {
      return content.trim();
    }
  }

  // Strategy 2: look for an interface/class entity that CONTAINS this method name
  // (the method is defined inside a class/interface body, with Javadoc)
  for (const c of allEntities) {
    if (c.entity_id === entity.entity_id) continue;
    if (c.file_path === entity.file_path) continue;
    if (c.entity_type !== "interface" && c.entity_type !== "class") continue;

    const content = c.after_content ?? c.before_content ?? "";
    if (!content.includes(entity.entity_name)) continue;
    // Only use interface entities, not other class implementations
    if (!content.includes("interface ")) continue;

    const contract = extractMethodContract(content, entity.entity_name);
    if (contract) return contract;
  }

  return null;
}

/**
 * Extract a method signature + preceding Javadoc from a class/interface body.
 */
function extractMethodContract(
  classContent: string,
  methodName: string,
): string | null {
  const lines = classContent.split("\n");
  for (let i = 0; i < lines.length; i++) {
    if (lines[i].includes(methodName) && (lines[i].includes("(") || lines[i].includes(";"))) {
      // Walk back to find Javadoc
      let start = i;
      for (let j = i - 1; j >= 0; j--) {
        const trimmed = lines[j].trim();
        if (trimmed === "" || trimmed.startsWith("*") || trimmed.startsWith("/**") || trimmed.startsWith("*/") || trimmed.startsWith("@")) {
          start = j;
        } else {
          break;
        }
      }
      // Include the signature line + Javadoc
      const contractLines = lines.slice(start, i + 1);
      const result = contractLines.map((l) => l.trimStart()).join("\n").trim();
      if (result.length > 10) return result;
    }
  }
  return null;
}

// Backward compat
export { SYSTEM_PROMPT as SYSTEM_PROMPT_FALLBACK };
export function buildFallbackPrompt(
  prTitle: string,
  diff: string,
  triageSection: string,
): string {
  return buildUserPrompt(prTitle, diff, triageSection, [], []);
}
